_examples/producer-confirms/producer.go

```diff
@@ -0,0 +1,241 @@
+	if err := channel.QueueBind(*routingKey, *routingKey, *exchange, false, nil); err != nil {
+		ErrLog.Fatalf("producer: Queue Bind: %s", err)
+	}
+
+	// Reliable publisher confirms require confirm.select support from the
+	// connection.
+	Log.Printf("producer: enabling publisher confirms.")
+	if err := channel.Confirm(false); err != nil {
+		ErrLog.Fatalf("producer: channel could not be put into confirm mode: %s", err)
+	}
+
+	for {
+		canPublish := false
+		Log.Println("producer: waiting on the OK to publish...")
+		for {
+			select {
+			case <-done_ch:
+				Log.Println("producer: stopping, all confirms seen")
+				return
+			case <-publish_ch:
+				Log.Println("producer: got the OK to publish")
+				canPublish = true
+				break
+			case <-time.After(time.Second):
+				WarnLog.Println("producer: still waiting on the OK to publish...")
+				continue
+			}
+			if canPublish {
+				break
+			}
+		}
+
+		Log.Printf("producer: publishing %dB body (%q)", len(*body), *body)
+		dConfirmation, err := channel.PublishWithDeferredConfirmWithContext(
+			ctx,
+			*exchange,
+			*routingKey,
+			true,
+			false,
+			amqp.Publishing{
+				Headers:         amqp.Table{},
+				ContentType:     "text/plain",
+				ContentEncoding: "",
+				DeliveryMode:    amqp.Persistent,
+				Priority:        0,
+				AppId:           "sequential-producer",
+				Body:            []byte(*body),
+			},
+		)
+		if err != nil {
+			ErrLog.Fatalf("producer: error in publish: %s", err)
+		}
+
+		select {
+		case <-done_ch:
+			Log.Println("producer: stopping, all confirms seen")
+			return
+		case confirms_ch <- dConfirmation:
+			Log.Println("producer: delivered deferred confirm to handler")
+			break
+		}
+
+		select {
+		case <-done_ch:
+			Log.Println("producer: stopping, all confirms seen")
+			return
+		case <-time.After(time.Second):
+			if *continuous {
+				continue
+			} else {
+				Log.Println("producer: initiating stop")
+				close(exit_ch)
+				select {
+				case <-done_ch:
+					Log.Println("producer: stopping, all confirms seen")
+					return
+				case <-time.After(time.Second * 10):
+					WarnLog.Println("producer: may be stopping with outstanding confirmations")
+					return
+				}
+			}
+		}
+	}
+}
+
+func startConfirmHandler(publish_ch chan<- struct{}, confirms_ch <-chan *amqp.DeferredConfirmation, done_ch chan struct{}, exit_ch <-chan struct{}) {
+	go func() {
+		confirms := make(map[uint64]*amqp.DeferredConfirmation)
+
+		for {
+			// Note: 8 is arbitrary, you may wish to allow more outstanding confirms before blocking publish
+			if len(confirms) <= 8 {
+				select {
+				case <-exit_ch:
+					exitConfirmHandler(confirms, done_ch)
+					return
+				default:
+					select {
+					case publish_ch <- struct{}{}:
+						Log.Println("confirm handler: sent OK to publish")
+					case <-time.After(time.Second * 5):
+						WarnLog.Println("confirm handler: timeout indicating OK to publish (this should never happen!)")
+					}
+				}
```

Comment:
Do we need the `default` case to then `select`?

I think we can just add the cases `case publish_ch <- struct{}{}` and `case <-time.After(time.Second * 5)` in the outter `select`, and it should behave the same?

Reply:
Ah yep, that was left over from when `publishCh` was unbuffered.